packages/mobile/src/exchange/saga.ts

```diff
@@ -1,5 +1,248 @@
-import { spawn, takeEvery, takeLatest } from 'redux-saga/effects'
-import { Actions, doFetchExchangeRate, exchangeGoldAndStableTokens } from 'src/exchange/actions'
+import { newKitFromWeb3 } from '@celo/contractkit'
+import {
+  ContractUtils,
+  getExchangeContract,
+  getGoldTokenContract,
+  getStableTokenContract,
+} from '@celo/walletkit'
+import { Exchange as ExchangeType } from '@celo/walletkit/types/Exchange'
+import { GoldToken as GoldTokenType } from '@celo/walletkit/types/GoldToken'
+import { StableToken as StableTokenType } from '@celo/walletkit/types/StableToken'
+import BigNumber from 'bignumber.js'
+import { call, put, select, spawn, takeEvery, takeLatest } from 'redux-saga/effects'
+import { showError } from 'src/alert/actions'
+import { ErrorMessages } from 'src/app/ErrorMessages'
+import { Actions, ExchangeTokensAction, setExchangeRate } from 'src/exchange/actions'
+import { ExchangeRatePair } from 'src/exchange/reducer'
+import { CURRENCY_ENUM } from 'src/geth/consts'
+import { RootState } from 'src/redux/reducers'
+import {
+  addStandbyTransaction,
+  generateStandbyTransactionId,
+  removeStandbyTransaction,
+} from 'src/transactions/actions'
+import { TransactionStatus, TransactionTypes } from 'src/transactions/reducer'
+import { sendAndMonitorTransaction } from 'src/transactions/saga'
+import { sendTransaction } from 'src/transactions/send'
+import { getRateForMakerToken, getTakerAmount } from 'src/utils/currencyExchange'
+import { roundDown } from 'src/utils/formatting'
+import Logger from 'src/utils/Logger'
+import { web3 } from 'src/web3/contracts'
+import { getConnectedAccount, getConnectedUnlockedAccount } from 'src/web3/saga'
+import * as util from 'util'
+
+const TAG = 'exchange/saga'
+
+const LARGE_DOLLARS_SELL_AMOUNT_IN_WEI = new BigNumber(1000 * 1000000000000000000) // To estimate exchange rate from exchange contract
+const LARGE_GOLD_SELL_AMOUNT_IN_WEI = new BigNumber(100 * 1000000000000000000)
+const EXCHANGE_DIFFERENCE_TOLERATED = 0.01 // Maximum difference between actual and displayed takerAmount
+
+export function* doFetchExchangeRate(makerAmount?: BigNumber, makerToken?: CURRENCY_ENUM) {
+  Logger.debug(TAG, 'Calling @doFetchExchangeRate')
+
+  // If makerAmount and makerToken are given, use them to estimate the exchange rate,
+  // as exchange rate depends on amount sold. Else default to present large sell amount.
+  const goldMakerAmount =
+    makerAmount && makerToken === CURRENCY_ENUM.GOLD ? makerAmount : LARGE_GOLD_SELL_AMOUNT_IN_WEI
+  const dollarMakerAmount =
+    makerAmount && makerToken === CURRENCY_ENUM.DOLLAR
+      ? makerAmount
+      : LARGE_DOLLARS_SELL_AMOUNT_IN_WEI
+
+  try {
+    yield call(getConnectedAccount)
+
+    const contractKit = newKitFromWeb3(web3)
```

Comment:
Should we have the kit live in web3/saga or somewhere else where more components can reach for it?